File: src/main/data/playerDataStore.ts
import { PlaylistDto, PlaylistItemDto, PlaylistListItemDto } from './playlistDto.js'
import { AppState } from './appState.js'
import Store from 'electron-store'
import path from 'path'
import { v4 as newUUID } from 'uuid'
import { PlayerRepository } from './playerRepository.js'

export class PlayerDataStore {
  playerRepository: PlayerRepository
  store: Store<AppState>

  constructor(userDataPath: string, migrationsPath: string) {
    this.playerRepository = new PlayerRepository(path.join(userDataPath, 'player.db'))
    this.initializeDatabase(migrationsPath)
    this.store = new Store<AppState>()
  }

  private initializeDatabase(migrationsPath: string): void {
    this.playerRepository.initializeDb(migrationsPath)
  }

  public GetCurrentPlaylistId(): string | undefined {
    return this.store.get('currentPlaylist')
  }

  public SetCurrentPlaylistId(id: string): void {
    this.store.set('currentPlaylist', id)
  }

  public async GetPlaylistById(id: string): Promise<PlaylistDto | undefined> {
    this.playerRepository.rebalancePlaylistOrder(id)
    const playlist = await this.playerRepository.getPlaylist(id)
    return playlist
  }

  public async GetPlaylistName(id: string): Promise<string | undefined> {
    const row = await this.playerRepository.getPlaylistProjection(id, { name: true })
    if (row === undefined) {
      return undefined
    }

    return row.name
  }

  public async GetPlaylists(): Promise<PlaylistListItemDto[]> {
    const items = await this.playerRepository.getPlaylists({
      id: true,
      name: true
    })

    return items.map(x => ({
      id: x.id,
      name: x.name
    }))
  }

  public async AddDefaultPlaylist() {
    if (!(await this.playerRepository.hasAnyPlaylist())) {
      const data = await this.AddPlaylist('default')
      this.SetCurrentPlaylistId(data.id)
    }
  }

  public async AddPlaylist(name: string) {
    const data = {
      id: newUUID(),
      name: name
    }
    await this.playerRepository.addPlaylist(data)

    return data
  }

  public UpdatePlaylistItem(id: string, data: { [x: string]: any }) {
    return this.playerRepository.updatePlaylistItem(id, data)
  }

  public UpdatePlaylistName(item: { id: string; name: string }): Promise<void> {
    return this.playerRepository.updatePlaylist(item.id, { name: item.name })
  }

  public UpdatePlaylist(id: string, data: { [x: string]: any }) {
    return this.playerRepository.updatePlaylist(id, data)
  }

  public InitNewPlaylistItems(filePaths: string[]): PlaylistItemDto[] {
    const result: PlaylistItemDto[] = []

    for (let i = 0; i < filePaths.length; i++) {
      const filePath = filePaths[i]
      result.push({
        id: newUUID(),
        path: filePath,
        title: path.basename(filePath),
        orderIndex: 0,
        isPlaying: null,
        currentTime: null,
        duration: null,
        progressPercent: null,
        lastWatched: null,
        playlistId: ''
      })
    }

    return result
  }

  public AddPlaylistItems(items: PlaylistItemDto[]) {
    this.playerRepository.addPlaylistItems(items.map(x => ({
      id: x.id,
      playlistId: x.playlistId,
      path: x.path,
      title: x.title,
      orderIndex: x.orderIndex
    })))
  }

  public DeletePlaylist(id: string) {
    return this.playerRepository.deletePlaylist(id)
  }

  public DeletePlaylistItem(id: string) {
    return this.playerRepository.deletePlaylistItem(id)
  }

  public UpdatePlaylistItemOrder(id: string, orderIndex: number) {
    return this.playerRepository.updatePlaylistItem(id, { orderIndex })
  }

  public RebalancePlaylistOrder(id: string) {
    this.playerRepository.rebalancePlaylistOrder(id)
  }
}

File: internal/data/player_data_store.go
package data

import (
	"fmt"
	"path/filepath"

	"github.com/google/uuid"
	graftikLogger "github.com/QuickOrBeDead/graftik-video-player/internal/logger"
)

type PlayerDataStore struct {
	repo   *PlayerRepository
	config *ConfigStore
	log    graftikLogger.Logger
}

type addItemInput struct {
	ID         string
	PlaylistID string
	Path       string
	Title      string
	OrderIndex float64
}

func NewPlayerDataStore(userDataPath, dbPath string, log graftikLogger.Logger) (*PlayerDataStore, error) {
	if log == nil {
		panic("data: logger is required")
	}
	log.Debug("data: creating player data store", "userDataPath", userDataPath, "dbPath", dbPath)
	repo, err := NewPlayerRepository(dbPath, log)
	if err != nil {
		return nil, err
	}
	config := NewConfigStore(userDataPath, log)
	if err := config.Load(); err != nil {
		return nil, fmt.Errorf("failed to load config: %w", err)
	}
	return &PlayerDataStore{repo: repo, config: config, log: log}, nil
}

func (s *PlayerDataStore) Initialize() error {
	s.log.Debug("data: initializing database")
	return s.repo.InitializeDB()
}

func (s *PlayerDataStore) Close() error {
	s.log.Debug("data: closing player data store")
	return s.repo.Close()
}

func (s *PlayerDataStore) GetCurrentPlaylistID() string {
	id := s.config.GetCurrentPlaylist()
	s.log.Debug("data: get current playlist id", "id", id)
	return id
}

func (s *PlayerDataStore) SetCurrentPlaylistID(id string) error {
	s.log.Debug("data: set current playlist id", "id", id)
	return s.config.SetCurrentPlaylist(id)
}

func (s *PlayerDataStore) GetPlaylistByID(id string) (*PlaylistDto, error) {
	s.log.Debug("data: getting playlist by id", "id", id)
	if err := s.repo.RebalancePlaylistOrder(id); err != nil {
		return nil, err
	}
	return s.repo.GetPlaylist(id)
}

func (s *PlayerDataStore) GetPlaylistName(id string) (string, error) {
	s.log.Debug("data: getting playlist name", "id", id)
	result, err := s.repo.GetPlaylistProjection(id, []string{"name"})
	if err != nil || result == nil {
		return "", err
	}
	name, _ := result["name"].(string)
	return name, nil
}

func (s *PlayerDataStore) GetPlaylists() ([]PlaylistListItem, error) {
	s.log.Debug("data: getting all playlists")
	return s.repo.GetPlaylists()
}

func (s *PlayerDataStore) AddDefaultPlaylist() error {
	s.log.Debug("data: adding default playlist")
	exists, err := s.repo.HasAnyPlaylist()
	if err != nil {
		return err
	}
	if !exists {
		id := uuid.New().String()
		if err := s.repo.AddPlaylist(id, "default"); err != nil {
			return err
		}
		return s.config.SetCurrentPlaylist(id)
	}

	// Playlists exist but no current playlist set — pick the first one
	if s.config.GetCurrentPlaylist() == "" {
		playlists, err := s.repo.GetPlaylists()
		if err != nil {
			return err
		}
		if len(playlists) > 0 {
			return s.config.SetCurrentPlaylist(playlists[0].ID)
		}
	}
	return nil
}

func (s *PlayerDataStore) AddPlaylist(name string) (*PlaylistDto, error) {
	s.log.Debug("data: adding playlist", "name", name)
	id := uuid.New().String()
	if err := s.repo.AddPlaylist(id, name); err != nil {
		return nil, err
	}
	return s.repo.GetPlaylist(id)
}

func (s *PlayerDataStore) UpdatePlaylist(id string, data map[string]any) error {
	s.log.Debug("data: updating playlist", "id", id)
	return s.repo.UpdatePlaylist(id, data)
}

func (s *PlayerDataStore) GetPlaylistItem(id string) *PlaylistItemDto {
	s.log.Debug("data: getting playlist item", "id", id)
	return s.repo.GetPlaylistItemByID(id)
}

func (s *PlayerDataStore) UpdatePlaylistItem(id string, data map[string]any) error {
	s.log.Debug("data: updating playlist item", "id", id)
	return s.repo.UpdatePlaylistItem(id, data)
}

func (s *PlayerDataStore) DeletePlaylist(id string) error {
	s.log.Debug("data: deleting playlist", "id", id)
	if err := s.repo.DeletePlaylist(id); err != nil {
		return err
	}
	if s.config.GetCurrentPlaylist() == id {
		return s.config.SetCurrentPlaylist("")
	}
	return nil
}

func (s *PlayerDataStore) DeletePlaylistItem(id string) error {
	s.log.Debug("data: deleting playlist item", "id", id)
	return s.repo.DeletePlaylistItem(id)
}

func (s *PlayerDataStore) InitNewPlaylistItems(filePaths []string) []PlaylistItemDto {
	s.log.Debug("data: initiating new playlist items", "count", len(filePaths))
	items := make([]PlaylistItemDto, len(filePaths))
	for i, fp := range filePaths {
		items[i] = PlaylistItemDto{
			ID:         uuid.New().String(),
			Path:       fp,
			Title:      filepath.Base(fp),
			OrderIndex: 0,
		}
	}
	return items
}

func (s *PlayerDataStore) AddPlaylistItems(items []PlaylistItemDto) {
	s.log.Debug("data: adding playlist items", "count", len(items))
	input := make([]struct {
		ID         string
		PlaylistID string
		Path       string
		Title      string
		OrderIndex float64
	}, len(items))
	for i, item := range items {
		input[i] = struct {
			ID         string
			PlaylistID string
			Path       string
			Title      string
			OrderIndex float64
		}{
			ID:         item.ID,
			PlaylistID: item.PlaylistID,
			Path:       item.Path,
			Title:      item.Title,
			OrderIndex: item.OrderIndex,
		}
	}
	s.repo.AddPlaylistItems(input)
}

func (s *PlayerDataStore) RebalancePlaylistOrder(id string) error {
	s.log.Debug("data: rebalancing playlist order", "id", id)
	return s.repo.RebalancePlaylistOrder(id)
}

func (s *PlayerDataStore) GetPreferences() *AppConfig {
	s.log.Debug("data: getting preferences")
	return s.config.GetPreferences()
}

func (s *PlayerDataStore) UpdateSettings(settings map[string]any) error {
	s.log.Debug("data: updating settings")
	return s.config.UpdateSettings(settings)
}
